File: project/backend/app/api/rooms.py
"""房间与成员接口 — 阶段 3.1，3.4 加历史/快照两个全员可读端点。

职责边界（照 cards.py 的模式）：
  - POST   /rooms            KP 建房：同一事务写 Room + KP 的 RoomMember
  - POST   /rooms/{id}/join  加入：校验房间/角色卡存在；同名重进幂等（不重复建行）
  - GET    /rooms            大厅列表（只列 waiting 状态的房间）
  - GET    /rooms/{id}       房间详情（退出时前端判断是否提示解散用）
  - DELETE /rooms/{id}       KP 解散房间（校验 kp_name，清成员与消息）
  - GET    /rooms/{id}/history  消息历史分页（3.4：暗骰行 D8 可见性过滤）
  - GET    /rooms/{id}/state    room_state 全量快照（3.4：重连补齐 / load 广播共用）

成员列表的组装逻辑抽成 _list_members()，join 的返回值与后续 WS 的
member_changed 广播共用同一份结构，避免两处手拼字段漂移。
"""
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlmodel import Session, select

from app.agent.kp_styles import style_echo
from app.db import get_session
from app.models import Card, Message, Room, RoomMember
from app.ws.manager import build_envelope, manager

router = APIRouter()


class RoomCreate(BaseModel):
    """建房请求体。KP 名即建房者昵称（无账号系统，名字就是身份）。"""

    name: str = Field(min_length=1, max_length=50)
    kp_name: str = Field(min_length=1, max_length=50)


class JoinRequest(BaseModel):
    """加入请求体。card_id 可空：允许先进房再看卡/建卡（3.3 侧栏再引导绑定）。"""

    player_name: str = Field(min_length=1, max_length=50)
    card_id: str | None = None


def _member_payload(m: RoomMember) -> dict:
    return {'player_name': m.player_name, 'role': m.role, 'card_id': m.card_id}


def _list_members(session: Session, room_id: str) -> list[dict]:
    """房间全量成员（REST 返回值与 WS member_changed 广播共用此结构）。"""
    rows = session.exec(
        select(RoomMember).where(RoomMember.room_id == room_id)
    ).all()
    return [_member_payload(m) for m in rows]


@router.post('/rooms', status_code=201)
def create_room(payload: RoomCreate, session: Session = Depends(get_session)):
    """KP 建房：Room 与 KP 成员记录同事务写入，保证有房必有 KP。"""
    room = Room(name=payload.name, kp_name=payload.kp_name)
    session.add(room)
    session.flush()  # 先拿生成的短码 id 再写成员
    session.add(RoomMember(room_id=room.id, player_name=payload.kp_name, role='kp'))
    session.commit()
    return {'room_id': room.id, 'name': room.name}


@router.post('/rooms/{room_id}/join')
def join_room(room_id: str, payload: JoinRequest, session: Session = Depends(get_session)):
    """加入房间：校验房间存在 / 角色卡存在；同名重进幂等，返回成员列表。

    重进语义（无账号系统，名字即身份）：
      - 报名 == kp_name：KP 重进，不新建成员行，身份保持 KP
      - 报名与现有玩家重名：视为同一人重连（刷新/误退重进），幂等返回花名册
    代价是"撞名顶号"（局域网熟人局可接受）；鉴权接入后此处收紧。
    """
    room = session.get(Room, room_id)
    if not room:
        raise HTTPException(status_code=404, detail='房间不存在')
    if payload.card_id and not session.get(Card, payload.card_id):
        raise HTTPException(status_code=404, detail='角色卡不存在')
    if payload.player_name == room.kp_name:
        # KP 重进幂等；但显式 leave（关网页 beacon）会删花名册行——行不在则补建，
        # 否则 KP 刷新后成员列表缺自己（4.4+ 修复）
        kp_row = session.exec(
            select(RoomMember).where(
                RoomMember.room_id == room_id,
                RoomMember.player_name == payload.player_name,
            )
        ).first()
        if not kp_row:
            session.add(RoomMember(
                room_id=room_id, player_name=payload.player_name, role='kp',
            ))
            session.commit()
        return _list_members(session, room_id)
    dup = session.exec(
        select(RoomMember).where(
            RoomMember.room_id == room_id,
            RoomMember.player_name == payload.player_name,
        )
    ).first()
    if dup:
        return _list_members(session, room_id)

    session.add(RoomMember(
        room_id=room_id,
        player_name=payload.player_name,
        role='player',
        card_id=payload.card_id,
    ))
    session.commit()
    return _list_members(session, room_id)


class LeaveRequest(BaseModel):
    """离开房间请求体（4.4+：关网页时 sendBeacon 发出）。"""

    player_name: str = Field(min_length=1, max_length=50)


@router.post('/rooms/{room_id}/leave')
async def leave_room(
    room_id: str, payload: LeaveRequest, session: Session = Depends(get_session),
):
    """显式离开房间（关网页 beacon / 退出按钮兜底）：删花名册行 + 系统消息 + 广播。

    4.4+ 语义：WS 断开一律按掉线处理，不再广播离开；只有显式 leave 才算退出。
    sendBeacon 在页面卸载时发不出 WS 帧，故走 REST。幂等：人不在花名册时静默返回。
    """
    room = session.get(Room, room_id)
    if not room:
        return {'ok': True}  # 房间已没了（解散后卸载），beacon 场景静默成功
    row = session.exec(
        select(RoomMember).where(
            RoomMember.room_id == room_id,
            RoomMember.player_name == payload.player_name,
        )
    ).first()
    if row is None:
        return {'ok': True}
    session.delete(row)
    session.add(Message(
        room_id=room_id, channel='system', type='sys',
        sender='system', content=f'{payload.player_name} 离开了房间',
    ))
    session.commit()
    members = _list_members(session, room_id)
    await manager.broadcast(room_id, build_envelope(
        'chat_new', room_id, 'system', 'system',
        {'text': f'{payload.player_name} 离开了房间'},
    ))
    await manager.broadcast(room_id, build_envelope(
        'member_changed', room_id, 'system', 'system', {'members': members},
    ))
    return {'ok': True}


@router.get('/rooms/{room_id}')
def get_room(room_id: str, session: Session = Depends(get_session)):
    """房间详情（轻量：不带成员，成员走 join/WS）。"""
    room = session.get(Room, room_id)
    if not room:
        raise HTTPException(status_code=404, detail='房间不存在')
    return {
        'room_id': room.id,
        'name': room.name,
        'kp_name': room.kp_name,
        'status': room.status,
        # 场景标题栏（3.3）：玩家进房即拿到当前场景
        'scene': {'scene_title': room.scene_title, 'scene_desc': room.scene_desc},
        # Agent 模式（4.1）：manual 纯人工 / collab 协同建议（KP 控制台面板回显用）
        'agent_mode': room.agent_mode,
        # KP 风格（4.4）：id + 名称（面板回显用）
        'style': style_echo(session, room),
    }


@router.get('/rooms')
def list_waiting_rooms(session: Session = Depends(get_session)):
    """大厅：只列等待中的房间（playing 的房间不重开，入口在 3.3 KP 控制台）。"""
    rows = session.exec(select(Room).where(Room.status == 'waiting')).all()
    return [
        {
            'room_id': r.id,
            'name': r.name,
            'kp_name': r.kp_name,
            'created_at': r.created_at.isoformat(),
        }
        for r in rows
    ]


@router.delete('/rooms/{room_id}', status_code=204)
async def dissolve_room(room_id: str, kp_name: str, session: Session = Depends(get_session)):
    """KP 解散房间：kp_name 对不上即 403（MVP 的轻量鉴权，名字即身份）。

    删除数据前先广播 room_dissolved 并关闭全房连接：在线成员收到信封即
    自行退出回大厅，遗留连接由 close_room 兜底关闭（4404=房间不存在），
    不再出现"人还在已解散的房间里发消息"的幽灵状态。
    """
    room = session.get(Room, room_id)
    if not room:
        raise HTTPException(status_code=404, detail='房间不存在')
    if kp_name != room.kp_name:
        raise HTTPException(status_code=403, detail='只有 KP 能解散房间')
    await manager.broadcast(
        room_id,
        build_envelope('room_dissolved', room_id, kp_name, 'system', {'operator': kp_name}),
    )
    await manager.close_room(room_id)
    for m in session.exec(select(RoomMember).where(RoomMember.room_id == room_id)).all():
        session.delete(m)
    for msg in session.exec(select(Message).where(Message.room_id == room_id)).all():
        session.delete(msg)
    session.delete(room)
    session.commit()


# ==================== 阶段 3.4：历史检索 + room_state 全量快照 ====================

def build_room_state(session: Session, room: Room) -> dict:
    """room_state 全量快照组装器：REST /state 与 load 的 WS 广播共用同一结构。

    members 用持久花名册全量语义（4.4+ 与 member_changed 广播一致：掉线成员
    仍在列表里，只有显式 leave 才删行；读档时花名册由快照重建，无幽灵行）；
    hp_sanity 按持久花名册全量组装——离线成员的 HP/SAN 也给全，重连补齐不缺行。
    HP/SAN 以上限 card_data 为准（投影列只作索引），上限取 derived 的 HP/SAN。
    """
    roster = session.exec(select(RoomMember).where(RoomMember.room_id == room.id)).all()

    members = [
        {'player_name': m.player_name, 'role': m.role, 'card_id': m.card_id}
        for m in roster
    ]

    hp_sanity: dict[str, dict] = {}
    for m in roster:
        if not m.card_id:
            continue
        card = session.get(Card, m.card_id)
        if not card:
            continue
        state = card.card_data.get('state', {})
        derived = card.card_data.get('derived', {})
        hp_sanity[m.player_name] = {
            'hp': int(state.get('current_hp', 0)),
            'sanity': int(state.get('current_sanity', 0)),
            'hp_max': int(derived.get('HP', 0)),
            'sanity_max': int(derived.get('SAN', 0)),
        }

    return {
        'members': members,
        'scene': {'scene_title': room.scene_title, 'scene_desc': room.scene_desc},
        'hp_sanity': hp_sanity,
        'style': style_echo(session, room),
    }


def _history_payload(rows: list[Message], roster_roles: dict[str, str]) -> list[dict]:
    """历史行序列化：payload/secret 原样透出（前端按 type 还原 ChatItem），
    role 由花名册现查（system 落库署名固定 'system'）。"""
    items = []
    for r in rows:
        items.append({
            'id': r.id,
            'channel': r.channel,
            'type': r.type,
            'sender': r.sender,
            'content': r.content,
            'secret': r.secret,
            'payload': r.payload,
            'role': 'system' if r.sender == 'system' else roster_roles.get(r.sender, 'player'),
            'created_at': r.created_at.isoformat(),
        })
    return items


@router.get('/rooms/{room_id}/history')
def get_room_history(
    room_id: str,
    viewer: str,
    before_id: int | None = None,
    limit: int = 100,
    session: Session = Depends(get_session),
):
    """消息历史分页（3.4）：按 id 倒序翻页，前端一次性拉全量后正序回放。

    D8 延伸（4.2 收紧为全类型）：secret 行（暗骰、AI 主持的 keeper 笔记）
    只对 viewer==kp_name 返回；其余行全员可见。
    """
    room = session.get(Room, room_id)
    if not room:
        raise HTTPException(status_code=404, detail='房间不存在')

    stmt = select(Message).where(Message.room_id == room_id)
    if before_id is not None:
        stmt = stmt.where(Message.id < before_id)
    if viewer != room.kp_name:
        stmt = stmt.where(Message.secret.is_(False))
    stmt = stmt.order_by(Message.id.desc()).limit(max(1, min(limit, 500)))

    rows = session.exec(stmt).all()
    roster_roles = {
        m.player_name: m.role
        for m in session.exec(select(RoomMember).where(RoomMember.room_id == room_id)).all()
    }
    return {'messages': _history_payload(list(rows), roster_roles)}


@router.get('/rooms/{room_id}/state')
def get_room_state(room_id: str, session: Session = Depends(get_session)):
    """room_state 全量快照（3.4）：重连成功后前端主动拉一次补齐状态。"""
    room = session.get(Room, room_id)
    if not room:
        raise HTTPException(status_code=404, detail='房间不存在')
    return build_room_state(session, room)

File: project/frontend/src/api/rooms.ts
import { client } from './client'
import type {
  AgentMode,
  CardStateSnapshot,
  RoomCreated,
  RoomListItem,
  RollResultPayload,
  StatusChangedPayload,
  WsMember,
} from '@/types/ws'

// 说明：响应拦截器已返回 response.data，统一用 axios 第二泛型取响应体。

/** KP 建房：返回 8 位短码房间号 */
export async function createRoom(name: string, kpName: string): Promise<RoomCreated> {
  return client.post<RoomCreated, RoomCreated>('/rooms', { name, kp_name: kpName })
}

/** 玩家加入房间：card_id 可空（先进房再绑卡），返回全量成员列表 */
export async function joinRoom(
  roomId: string,
  playerName: string,
  cardId: string | null,
): Promise<WsMember[]> {
  return client.post<WsMember[], WsMember[]>(`/rooms/${roomId}/join`, {
    player_name: playerName,
    card_id: cardId,
  })
}

/** 大厅：等待中的房间列表 */
export async function listRooms(): Promise<RoomListItem[]> {
  return client.get<RoomListItem[], RoomListItem[]>('/rooms')
}

/** 房间详情（轻量）：退出房间时判断是否提示 KP 解散用；scene 供进房恢复场景标题栏；
 * agent_mode（4.1）供 KP 控制台 AI 建议面板回显；style（4.4）供 KP 风格面板回显 */
export interface RoomDetail {
  room_id: string
  name: string
  kp_name: string
  status: 'waiting' | 'playing'
  scene: { scene_title: string; scene_desc: string }
  agent_mode: AgentMode
  style: { style_id: string; style_name: string }
}

export async function getRoom(roomId: string): Promise<RoomDetail> {
  return client.get<RoomDetail, RoomDetail>(`/rooms/${roomId}`)
}

/** KP 解散房间（kp_name 校验失败返回 403，由拦截器统一提示） */
export async function dissolveRoom(roomId: string, kpName: string): Promise<void> {
  await client.delete(`/rooms/${roomId}`, { params: { kp_name: kpName } })
}

// ==================== 阶段 3.3：KP 控制台 ====================

/** POST /rooms/{id}/status 响应（clamp 后的绝对值） */
export interface StatusResult extends CardStateSnapshot {
  target: string
}

/** KP 改成员 HP/SAN：hp/sanity 传"变更后的绝对值"（服务端 clamp 到 [0, derived]），
 * reason 必填（D9：一切数值变更必须可溯源） */
export interface RoomStatusChange {
  kp_name: string
  target: string
  hp?: number
  sanity?: number
  reason: string
}

export async function updateRoomStatus(
  roomId: string,
  body: RoomStatusChange,
): Promise<StatusResult> {
  return client.post<StatusResult, StatusResult>(`/rooms/${roomId}/status`, body)
}

/** KP 更新场景标题栏（本地 scene 由 scene_changed 广播回填，不在调用处自改） */
export async function updateRoomScene(
  roomId: string,
  body: { kp_name: string; scene_title: string; scene_desc?: string },
): Promise<void> {
  await client.put(`/rooms/${roomId}/scene`, body)
}

/** KP 宣布开团：waiting → playing（幂等；开团后大厅不再展示本房间） */
export async function startGame(
  roomId: string,
  kpName: string,
): Promise<{ status: 'playing'; already_started: boolean }> {
  return client.put(`/rooms/${roomId}/start`, { kp_name: kpName })
}

// ==================== 阶段 3.4：历史 / 快照 / 存读档 ====================

/** room_state 快照（GET /rooms/{id}/state 响应；WS room_state 信封 payload 同构） */
export interface RoomStateSnapshot {
  members: WsMember[]
  scene: { scene_title: string; scene_desc: string }
  hp_sanity: Record<string, CardStateSnapshot>
  style?: { style_id: string; style_name: string }
}

/** history 接口的一行（与后端 _history_payload 一一对应）。
 * payload 是落库时的结构化原始数据：dice 行为 RollResultPayload、
 * status 行为 StatusChangedPayload、text 行为 {role}；旧行/异常行可能为 null */
export interface HistoryMessage {
  id: number
  channel: 'narrative' | 'ooc' | 'system'
  type: 'text' | 'dice' | 'sys' | 'status' | 'check_request'
  sender: string
  content: string
  secret: boolean
  payload: Record<string, any> | null
  role: 'kp' | 'player' | 'system'
  created_at: string
}

/** 消息历史（id 倒序返回，前端回放前 reverse）。暗骰行服务端已按 viewer 过滤 */
export interface RoomHistory {
  messages: HistoryMessage[]
}

/** 拉历史：一次拉全量（limit 500），翻页接口有 before_id 但 MVP 不做无限滚动 */
export async function getRoomHistory(
  roomId: string,
  viewer: string,
  limit = 500,
): Promise<RoomHistory> {
  return client.get<RoomHistory, RoomHistory>(`/rooms/${roomId}/history`, {
    params: { viewer, limit },
  })
}

/** room_state 全量快照：重连成功后主动拉一次补齐成员/场景/HP·SAN */
export async function getRoomState(roomId: string): Promise<RoomStateSnapshot> {
  return client.get<RoomStateSnapshot, RoomStateSnapshot>(`/rooms/${roomId}/state`)
}

/** POST /rooms/{id}/save 响应 */
export interface SaveCreated {
  save_id: number
  name: string
  created_at: string
}

/** KP 存档：快照 = 成员花名册 + 各卡完整 card_data + 场景 */
export async function saveGame(roomId: string, kpName: string, name: string): Promise<SaveCreated> {
  return client.post<SaveCreated, SaveCreated>(`/rooms/${roomId}/save`, {
    kp_name: kpName,
    name,
  })
}

/** GET /rooms/{id}/saves 列表项 */
export interface SaveMeta {
  id: number
  name: string
  created_at: string
}

export async function listSaves(roomId: string): Promise<SaveMeta[]> {
  return client.get<SaveMeta[], SaveMeta[]>(`/rooms/${roomId}/saves`)
}

/** KP 读档：快照原子写回，返回 room_state 快照；界面由 room_state 广播驱动 */
export async function loadSave(
  roomId: string,
  saveId: number,
  kpName: string,
): Promise<RoomStateSnapshot> {
  return client.post<RoomStateSnapshot, RoomStateSnapshot>(
    `/rooms/${roomId}/load/${saveId}`,
    { kp_name: kpName },
  )
}

// ==================== 4.3+：检定下放 ====================

/** KP 手动发起检定下放（collab/manual 模式的「要求玩家投骰」；auto 由 AI 发起）。
 * 服务端按目标角色卡查技能值（D5），落 check_request 消息并 chat_new 广播 */
export async function createCheckRequest(
  roomId: string,
  body: {
    kp_name: string
    target: string
    skill_name: string
    difficulty: 'standard' | 'hard' | 'extreme'
    reason: string
  },
): Promise<{ status: string; request_id: string; target: string }> {
  return client.post(`/rooms/${roomId}/check-requests`, body)
}

/** 被点名的玩家投掷检定：服务端按请求里存好的技能值判定（D5，玩家无自掷空间）。
 * 结算结果同时落 type=dice 行 + roll_result 广播；auto 模式下唤醒 AI 继续 */
export async function rollCheckRequest(
  roomId: string,
  requestId: string,
  playerName: string,
): Promise<RollResultPayload> {
  return client.post<RollResultPayload, RollResultPayload>(
    `/rooms/${roomId}/check-requests/${requestId}/roll`,
    { player_name: playerName },
  )
}
